### PR TITLE
feat: RepeatedTask adds execute-first-wait-later behavior.

### PR DESCRIPTION
**Pull request by paomian**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

RepeatedTask adds execute-first-wait-later behavior.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/common/runtime/src/repeated_task.rs [hunk 1]
```diff
@@ -42,14 +43,69 @@ struct TaskInner<E> {
...
     task_fn: Option<BoxedTaskFunction<E>>,
+    /// Generates the next interval.
+    interval_generator: Option<Box<dyn IntervalGenerator>>,
+}
+
+pub trait IntervalGenerator: Send + Sync {
+    fn next(&mut self) -> Duration;
+    fn is_regular(&self) -> bool {
+        true
+    }
```
Comment (WenyXu) [inline src/common/runtime/src/repeated_task.rs:54]:
```suggestion
    fn next(&mut self) -> Duration;
    
    fn is_regular(&self) -> bool {
        true
    }
```

File: src/common/runtime/src/repeated_task.rs [hunk 2]
```diff
@@ -42,14 +43,69 @@ struct TaskInner<E> {
...
+}
+
+impl From<Duration> for Box<dyn IntervalGenerator> {
+    fn from(value: Duration) -> Self {
+        Box::new(value)
+    }
+}
+
+pub struct ImmediatelyInterval {
```
Comment (WenyXu) [inline src/common/runtime/src/repeated_task.rs:69]: How about `FirstZeroInterval`, I asked the OpenAI 🤣

File: src/common/runtime/src/repeated_task.rs (hunk 1 above)
Comment (WenyXu) [reply src/common/runtime/src/repeated_task.rs:54]: BTW, How about allowing the `IntervalGenerator` to return self's display info?
Such as: 

```
pub trait IntervalGenerator: Send + Sync {
    fn next(&mut self) -> Duration;

    fn display(&self) -> String;
```

**Review by WenyXu — APPROVED** [review]

File: src/common/runtime/src/repeated_task.rs [hunk 3]
```diff
@@ -40,16 +41,89 @@ pub type BoxedTaskFunction<E> = Box<dyn TaskFunction<E> + Send + Sync + 'static>
...
+
+    /// Generates the next interval.
+    interval_generator: Option<Box<dyn IntervalGenerator>>,
+}
+
+pub trait IntervalGenerator: Send + Sync {
+    /// return the next interval.
+    fn next(&mut self) -> Duration;
+
+    /// return whether the interval is regular and the interval if it is regular.
```
Comment (WenyXu) [inline src/common/runtime/src/repeated_task.rs:56]: nit: 
```suggestion
    /// Returns whether the interval is regular and the interval if it is regular.
```

File: src/common/runtime/src/repeated_task.rs [hunk 4]
```diff
@@ -40,16 +41,89 @@ pub type BoxedTaskFunction<E> = Box<dyn TaskFunction<E> + Send + Sync + 'static>
...
+
     /// The task_fn to run. This is Some if the task is not started.
     task_fn: Option<BoxedTaskFunction<E>>,
+
+    /// Generates the next interval.
+    interval_generator: Option<Box<dyn IntervalGenerator>>,
+}
+
+pub trait IntervalGenerator: Send + Sync {
+    /// return the next interval.
```
Comment (WenyXu) [inline src/common/runtime/src/repeated_task.rs:53]: nit:
```suggestion
    /// Returns the next interval.
```

File: src/common/runtime/src/repeated_task.rs [hunk 5]
```diff
@@ -75,16 +149,19 @@ impl<E> Drop for RepeatedTask<E> {
 }
 
 impl<E: ErrorExt + 'static> RepeatedTask<E> {
-    pub fn new(interval: Duration, task_fn: BoxedTaskFunction<E>) -> Self {
+    pub fn new<I: Into<Box<dyn IntervalGenerator>>>(
```
Comment (evenyag) [inline src/common/runtime/src/repeated_task.rs:152]: I think this makes things complicated. In most scenarios, what we need is an initial delay. Now our initial delay is always `interval`. We can add an argument to `new()` or provide another method to specify the initial delay:
```rust
pub fn with_initial_delay(initial_delay: Duration, interval: Duration, task_f: BoxedTaskFunction<E>) -> Self {}
```

Then we can call the task_fn first. By default, `initial_delay` is None.
```rust
loop {
    let sleep_time = initial_delay.take().unwrap_or(interval);
    tokio::select! {}
}
```

File: src/common/runtime/src/repeated_task.rs [hunk 6]
```diff
@@ -85,9 +87,15 @@ impl<E: ErrorExt + 'static> RepeatedTask<E> {
...
             started: AtomicBool::new(false),
             interval,
+            initial_delay: None,
         }
     }
 
+    pub fn with_initial_delay(mut self, initial_delay: Duration) -> Self {
+        self.initial_delay = Some(initial_delay);
+        self
+    }
```
Comment (killme2008) [inline src/common/runtime/src/repeated_task.rs:97]:
```suggestion
    pub fn with_initial_delay(mut self, initial_delay: Option<Duration>) -> Self {
        self.initial_delay = initial_delay;
        self
    }
```

File: src/common/runtime/src/repeated_task.rs [hunk 7]
```diff
@@ -99,15 +107,17 @@ impl<E: ErrorExt + 'static> RepeatedTask<E> {
...
-        let interval = self.interval;
         let child = self.cancel_token.child_token();
         // Safety: The task is not started.
         let mut task_fn = inner.task_fn.take().unwrap();
+        let interval = self.interval;
+        let mut initial_delay = self.initial_delay;
         // TODO(hl): Maybe spawn to a blocking runtime.
         let handle = runtime.spawn(async move {
             loop {
+                let sleep_time = initial_delay.take().unwrap_or(interval);
```
Comment (killme2008) [inline src/common/runtime/src/repeated_task.rs:118]: What about checking if the `sleep_time` is greater than zero? If not, skip the `tokio::select!`? Avoid the invocation.

File: src/common/runtime/src/repeated_task.rs [hunk 8]
```diff
@@ -50,6 +51,7 @@ pub struct RepeatedTask<E> {
     inner: Mutex<TaskInner<E>>,
     started: AtomicBool,
     interval: Duration,
+    initial_delay: Option<Duration>,
```
Comment (evenyag) [inline src/common/runtime/src/repeated_task.rs:54]: Document the initial delay is `interval` in `new()`.

**Review by evenyag — APPROVED** [review]
Rest LGTM

File: src/common/runtime/src/repeated_task.rs (hunk 7 above)
Comment (paomian) [reply src/common/runtime/src/repeated_task.rs:118]: good idea

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**